File: CCC/ccc09s5.cpp
#include <bits/stdc++.h>
using namespace std;
int M, N, K, a[30002][1002];
int main(){
    scanf("%d%d%d",&M,&N,&K);
    int x,y,r,b;
    while(K--){
        scanf("%d%d%d%d",&x,&y,&r,&b);
        x--, y--;
        for(int i = max(0, x-r),dif; i <= min(N-1, x+r); i++){
            dif = (int)sqrt(r*r - (x-i)*(x-i));
            a[max(0, y-dif)][i] += b;
            if(y+dif+1 < M)
                a[y+dif+1][i] -= b;
        }
    }
    int best = 0, cnt = 0;
    
    for(int i = 0,sum; i < N; i++){
        sum = 0;
        for(int j = 0; j < M; j++){
            sum += a[j][i];
            if(sum > best){
                best = sum;
                cnt = 1;
            }
            else if(sum == best)
                cnt++;
        }
    }
    printf("%d\n%d\n", best, cnt);
    return 0;
}

File: CCC/ccc09s5.java
import java.util.*;
import java.io.*;
public class ccc09s5{
	static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
	static StringTokenizer st;
	
	public static void main(String[] args) throws IOException{
		int m = readInt();
		int n = readInt();
		int k = readInt();

		int[][] a = new int[n+1][m+1];
		while(k --> 0){
			int x = readInt();
			int y = readInt();
			int r = readInt();
			int b = readInt();
			for(int i = Math.max(1, x-r); i <= Math.min(n, x+r); i++){
				int d = (int)Math.sqrt(r*r - (x-i)*(x-i));
				a[i][Math.max(1, y-d)] += b;
				if(y+d+1 <= m)
					a[i][y+d+1] -= b;
			}
		}
		int best = 0, cnt = 0;
		for(int i = 1; i <= n; i++){
			for(int j = 1; j <= m; j++){
				a[i][j] += a[i][j-1];
				if(a[i][j] > best){
					best = a[i][j];
					cnt = 1;
				}
				else if(a[i][j] == best)
					cnt++;
			}
		}
		System.out.println(best);
		System.out.println(cnt);
	}


	static String next() throws IOException{
		while (st == null || !st.hasMoreTokens())
			st = new StringTokenizer(br.readLine().trim());
		return st.nextToken();
	}
	static long readLong() throws IOException{
		return Long.parseLong(next());
	}
	static int readInt() throws IOException{
		return Integer.parseInt(next());
	}
	static double readDouble() throws IOException{
		return Double.parseDouble(next());
	}
	static char readCharacter() throws IOException{
		return next().charAt(0);
	}
	static String readLine() throws IOException{
		return br.readLine().trim();
	}
}